packages/aragon-wrapper/src/index.js

```diff
@@ -348,59 +287,99 @@ export default class Aragon {
    * @return {void}
    */
   initApps () {
-    // TODO: Refactor this a bit because it's pretty much an eye sore
-    this.identifiers = new Subject()
-    this.appsWithoutIdentifiers = this.permissions.pipe(
+    // Cache requests so we don't make unnecessary calls when a call is already in-flight
+    const applicationInfoCache = new AsyncRequestCache((cacheKey) => {
+      const [appId, codeAddress] = cacheKey.split('.')
+      return getAragonOsInternalAppInfo(appId) ||
+        this.apm.getLatestVersionForContract(appId, codeAddress)
+    })
+
+    const proxyContractValueCache = new AsyncRequestCache((proxyAddress) => {
+      if (this.isKernelAddress(proxyAddress)) {
+        const kernelProxy = makeProxy(proxyAddress, 'ERCProxy', this.web3, this.kernelProxy.initializationBlock)
+
+        return Promise.all([
+          // Use Kernel ABI
+          this.kernelProxy.call('KERNEL_APP_ID'),
+          // Use ERC897 proxy ABI
+          // Note that this won't work on old Aragon Core 0.5 Kernels,
+          // as they had not implemented ERC897 yet
+          kernelProxy.call('implementation')
+        ]).then((values) => ({
+          appId: values[0],
+          codeAddress: values[1]
+        }))
+      }
+
+      const appProxy = makeProxy(proxyAddress, 'AppProxy', this.web3, this.kernelProxy.initializationBlock)
+      const appProxyForwarder = makeProxy(proxyAddress, 'Forwarder', this.web3, this.kernelProxy.initializationBlock)
+
+      return Promise.all([
+        appProxy.call('kernel'),
+        appProxy.call('appId'),
+        appProxy.call('implementation'),
+        // Not all apps implement the forwarding interface
+        appProxyForwarder.call('isForwarder').catch(() => false)
+      ]).then((values) => ({
+        kernelAddress: values[0],
+        appId: values[1],
+        codeAddress: values[2],
+        isForwarder: values[3]
+      }))
+    })
+
+    // Get all app proxy addresses
+    const baseApps$ = this.permissions.pipe(
       map(Object.keys),
       // Add Kernel as the first "app"
-      map((addresses) => {
-        const appsWithoutKernel = addresses.filter((address) => !this.isKernelAddress(address))
+      map((proxyAddresses) => {
+        const appsWithoutKernel = proxyAddresses.filter((address) => !this.isKernelAddress(address))
         return [this.kernelProxy.address].concat(appsWithoutKernel)
       }),
       // Get proxy values
       switchMap(
-        (appAddresses) => Promise.all(
-          appAddresses.map((app) => this.getProxyValues(app))
+        (proxyAddresses) => Promise.all(
+          proxyAddresses.map(async (proxyAddress) => {
+            let proxyValues
```

Comment:
Can we simplify this step by adding the `proxyAddress` to the object constructed in the `requestFn` of `proxyContractValueCache`?

Reply:
I like the current implementation a bit better, as it'll ensure `proxyAddress` is added even if the request fails. We'd otherwise always have to resolve in `proxyContractValueCache`, which would break the cache invalidation mechanism on fails.